tensorflow/core/kernels/stack.cc

```diff
@@ -319,6 +413,40 @@ void StackPopOp::ComputeAsync(OpKernelContext* ctx, DoneCallback done) {
     ctx->set_output(0, value.tensor);
     done();
   }
+
+  // This is the reverse of the heuristic for swapping out, asynchronously
```

Comment:
The only difference from the original `StackPopOp` is from here and on.

Apart from the current tensor, it asynchronously swaps back the latest swapped tensor, so that when it's needed it's already swapped in.

This is the core optimization idea. Since the swapped tensors are the oldest, more tensors can be asynchronously swapped in increasing the overlapping between computations and swapping, thus reducing the overall training time.